packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -59,48 +93,58 @@ import 'transitions.dart';
 ///   // here `bytes` is a Uint8List containing the bytes for the in-memory image
 ///   placeholder: MemoryImage(bytes),
 ///   image: NetworkImage('https://backend.example.com/image.png'),
+///   animationBuilder: (Widget target, Widget placeholder, bool isTargetLoaded) => AnimatedFadeOutFadeIn(
+///     target: target,
+///     placeholder: placeholder,
+///     isTargetLoaded: isTargetLoaded,
+///     fadeInDuration: const Duration(milliseconds: 300),
+///     fadeOutDuration: Curves.easeOut,
+///     fadeInCurve: const Duration(milliseconds: 700),
+///     fadeOutCurve: Curves.easeIn,
+///   );
 /// )
 /// ```
 /// {@end-tool}
 class FadeInImage extends StatelessWidget {
   /// Creates a widget that displays a [placeholder] while an [image] is loading,
-  /// then fades-out the placeholder and fades-in the image.
+  /// then fades from the placeholder to the image using the [animationBuilder].
   ///
   /// The [placeholder] and [image] may be composed in a [ResizeImage] to provide
   /// a custom decode/cache size.
   ///
-  /// The [placeholder], [image], [fadeOutDuration], [fadeOutCurve],
-  /// [fadeInDuration], [fadeInCurve], [alignment], [repeat], and
-  /// [matchTextDirection] arguments must not be null.
+  /// The [placeholder], [image], [alignment], [repeat], and [matchTextDirection]
+  /// arguments must not be null.
+  /// 
+  /// If the [animationBuilder] argument is null then the [fadeOutDuration],
+  /// [fadeOutCurve], [fadeInDuration] and [fadeInCurve] arguments must not be
+  /// null.
   ///
   /// If [excludeFromSemantics] is true, then [imageSemanticLabel] will be ignored.
-  const FadeInImage({
+  FadeInImage({
```

Comment:
We shouldn't change the API of FadeInImage. This seems like an unnecessary breaking change. Instead, if we find an appropriate abstraction, FadeInImage should keep its current functionality and under the hood just implement this in terms of a to-be-created widget that allows the configuration of the placeholder->image transition.

Reply:
I agree, I couldn't work out how to keep this constructor `const` while also setting up a default `animationBuilder` in the constructor. Reworked this to instead fallback during the `build` cycle.